flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/PartitionFetcher.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.filesystem;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.java.tuple.Tuple2;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * Fetcher to fetch the suitable partitions of a filesystem table.
+ *
+ * @param <P> The type to describe a partition.
+ */
+@Internal
+public interface PartitionFetcher<P> extends Serializable {
+
+	/**
+	 * Open the resources of the fetcher.
+	 */
+	void open() throws Exception;
+
+	/**
+	 * Fetch the suitable partitions, call this method should guarantee the fetcher has opened.
+	 */
+	List<P> fetch() throws Exception;
```

Comment:
I think it is better to keep `Context` here, and we don't need `open` and `close`.
(All `open` and `close` are the same)